galaxy/worker/tasks/collection.py

```diff
@@ -0,0 +1,143 @@
+                     .format(task.id, artifact_id))
+    try:
+        collection_info = _process_collection(
+            artifact, filename, task_logger)
+        _publish_collection(task, artifact, repository, collection_info)
+    except exc.TaskError as e:
+        artifact.delete()
+        task_logger.error('Import Task "{task_id}" failed: {message}'
+                          .format(task_id=task.id, message=str(e)))
+        raise
+
+    warnings, errors = task.get_message_stats()
+    msg = ('Import completed with {warnings} warnings and {errors} errors'
+           .format(warnings=warnings, errors=errors))
+    task_logger.info(msg)
+
+
+def _get_task_logger(task):
+    logger = logging.getLogger('galaxy.worker.tasks.import_collection')
+    return logutils.ImportTaskAdapter(logger, task=task)
+
+
+def _process_collection(artifact, filename, task_logger):
+    with tempfile.TemporaryDirectory() as extract_dir:
+        with artifact.file.open() as pkg_file, \
+                tarfile.open(fileobj=pkg_file) as pkg_tar:
+            pkg_tar.extractall(extract_dir)
+
+        try:
+            collection_info = importer.import_collection(
+                extract_dir, filename, task_logger)
+        except i_exc.ImporterError as e:
+            raise exc.ImportFailed(str(e))
+
+    _log_collection_info(collection_info)
+    return collection_info
+
+
+@transaction.atomic
+def _publish_collection(task, artifact, repository, collection_info):
+    metadata = collection_info.collection_info
+    collection, _ = models.Collection.objects.get_or_create(
+        namespace=task.namespace, name=metadata.name)
+
+    try:
+        version = collection.versions.create(
+            version=metadata.version,
+            metadata=metadata.get_json(),
+            quality_score=collection_info.quality_score,
+        )
+    except IntegrityError:
+        raise exc.VersionConflict('Collection version already exists.')
+
+    relative_path = '{0}-{1}-{2}.tar.gz'.format(
+        metadata.namespace,
+        metadata.name,
+        metadata.version
+    )
+    pulp_models.ContentArtifact.objects.create(
+        artifact=artifact,
+        content=version,
+        relative_path=relative_path,
+    )
+
+    with pulp_models.RepositoryVersion.create(repository) as new_version:
+        new_version.add_content(
+            pulp_models.Content.objects.filter(pk=version.pk)
+        )
+
+    publication = pulp_models.Publication.objects.create(
+        repository_version=new_version,
+        complete=True,
+        pass_through=True,
+    )
+    pulp_models.Distribution.objects.update_or_create(
+        name='galaxy',
+        base_path='galaxy',
+        defaults={'publication': publication},
+    )
+
+    task.imported_version = version
+    task.save()
+
+
+def _log_collection_info(collection_info):
+    log.debug('Collection loaded - metadata={}, quality_score={}'.format(
+        collection_info.collection_info.__dict__,
```

Comment:
I would be little hesitant to use `collection_info.__dict__` in the string formatting here, just because __dict__ could potentially end up with values that have heavy or finicky str/repr's. 

I think with this style of string formatting the formatting is done before the log.debug() call, the str/repr's for collection_info.__dict__ will always run, even for log levels below DEBUG.

As opposed to something like:

``` python
log.debug('Collection loaded - metadata=%s, quality_score=%s', collection_info.__dict__, collection_info.quality_score)
```

But if the goal is to render the string message now (as opposed to some later point in the logging stack), I think the code as is works.

Reply:
These `log.debug` messages were added just for debugging locally and analysis on galaxy-dev, so using %-style strings would be fine